Implement start local and remote processes in Node

squirrelsc opened the pull request:
1. Implement start local and remote processes in Node.
1. Implement tools: echo, uname. They don't need installation, will support complex tools later.
1. update settings of vscode.
1. add description filed for test suite and case.
1. change to a better singleton pattern.
1. use same path as config file, so that lisa don't need to be start from code folder.
1. remove common folder, all things get into util folder.
1. refactoring code to use f format
1. fix most mypy errors.
1. rename test suite methods for better intuition.
1. other minor changes.

squirrelsc committed:
fix most isort errors.

squirrelsc committed:
change to a better singleton pattern.

squirrelsc committed:
use base path from config file

it's easier to put all things together.

squirrelsc committed:
implement local and remote processes on node

remove common folder
add some cleanup code

squirrelsc committed:
fix flake8

andyleejordan reviewed (changes requested):
Thank you Chi for your very quick prototype work. It's a good start. I think we need to pause new development and think through (and document) all the design choices made, and the existing implementations. We need to be certain that our base (this prototype) is exactly how we want it before we start productizing it.

squirrelsc committed:
use f format to replace %

1. rename test suite methods for better intuition.
2. add description for suite and case.
3. other minor changes

squirrelsc committed:
remove useless 'if' and use pathlib

squirrelsc committed:
fix by review comments.

squirrelsc committed:
implement simple tools: echo, uname

andyleejordan committed:
Fix bug in process.py

While the documentation recommends that the `subprocess` module is used
via the `run()` interface, we are currently using the underlying `Popen`
interface. The documentation further recommends that if
`shell=True` (which is currently the case in our code), that the command
is passed directly as a string, not a split argument sequence. Using
this recommendation fixed the existing bug where the arguments were
being ignored on Linux (causing `uname` to print “Linux” and not the
kernel version, release, and hardware info). However, we need to revisit
our wrapping of `subprocess` because it is difficult to get right.

andyleejordan committed:
Fix type errors

andyleejordan committed:
Simplify `_load_extends`

andyleejordan committed:
Simplify `warn_as_error`

andyleejordan committed:
Hide class and IDs of `ActionStatus` enum

Using the Enum module’s functional API we can hide the implementation
details that it is technically a class and that the members are
internally represented as integers, providing a clearer abstraction.

andyleejordan committed:
Rename `cleanup` to `close`

Which is Python’s canonical term for this logic. We ought to make as
many of these into context managers as possible, but for now let’s at
least use the expected names.

andyleejordan committed:
Fix misspelling of `executableResult`

I couldn’t find the file.

andyleejordan committed:
Use `dataclass` for `ExecutableResult`

Also clean the stdout/stderr at its source, not when saving. Use
`strip()` with its default arguments (which strips surrounding
whitespace, not just carriage returns and newlines).

andyleejordan committed:
Remove `Optional` from `config.py` to simplify types

And note our hack which indicates we need a structured object.

andyleejordan reviewed (approved):
There's of course more to consider and review (as noted), but I think we can merge this now.
